### DISCUSSION TITLE
Is there a good way of debugging code that uses the Xmodem transfers?

### DISCUSSION THREAD

**Discussion by rondnelson99** [TomNisbet/TommyPROM#29] [Q&A] [answered]
Hi, 
Thanks for putting this project out there, I've used it quite a bit for various purposes and it's certainly very good at what it does.
Recently I've been trying to use your Xmodem code in another project that basically just takes a file from a computer and sends it out over SPI in a custom protocol. Things haven't gone super smoothly due to my C++ abilities being rather lacking, and the project requiring me to use interrupts since the Arduino is an SPI slave. 

Anyways, I got to the point where the transfers *seemed* to be working, but the final output was all 0xFFs. Clearly something's wrong, and I wish I could use Serial.println to get an idea of what's going on. Of course the Serial port on the arduino is busy with the Xmodem transfer. If I hit cancel in Tera Term, the Arduino seems to just reset. (maybe that's part of the problem?) Is there an easy way to debug things while an Xmodem transfer is active? How did you debug your Xmodem code?
Thanks for your great code,
Ron

**Comment by TomNisbet** [comment, marked as answer]
Ron,

I did most of the debugging by just storing a string that could be printed after the transfer finished.  That's what the CmdStatus code does.  

Some additional variables were added later that are printed with the INFO command.  See PromDevice.h and cpp for the debugXXX variables.  You can do a lot with this approach: store packet numbers to see how far you are getting, store the first character of the packet, etc.  You can even store one or more debug strings and print them later.  Maybe build a small buffer, like 16 characters, and add code that copies the beginning bytes of the packet to it when sequencenumber==1.  Then you can dump this after the transfer and compare it to the start of your file.

If you really need live output, define a second serial port using the SoftwareSerial library and get a TTL to USB serial adapter to hook it up.  You will need to turn the port speeds of your XModem port way down so that the debug port has enough time to keep up.  I suspect you can do all of the debugging you need by just storing variables though.

If Teraterm shows that the transfers are proceeding then that means your code is seeing the data from the host well enough to pull out the packet numbers and checksums to form its response.  That would seem to indicate that the problem is either the code that is copying the data out of the packets or the code further down the line that is sending it out.


Good luck with your project.  Let me know if you need any additional information.
Tom

**Reply by rondnelson99** [reply]
Thanks! That gives me a lot to go on. I assumed that the actual Xmodem code was fine, but I hadn't thought of just storing data to be printed away until needed. I'll get back to you if I have any more questions.
